ygen/gogen.go

```diff
@@ -1296,48 +1299,48 @@ func writeGoStruct(targetStruct *Directory, goStructElements map[string]*Directo
 			zeroValue := mtype.ZeroValue
 			defaultValue := goLeafDefault(field, mtype)
 
-			if len(mtype.UnionTypes) > 1 {
-				// If this is a union that has more than one subtype, then we need
-				// to generate the relevant interface.
-				if _, ok := state.generatedUnionInterfaces[targetStruct.Name][mtype.NativeType]; !ok {
-					// If the union type has not already been generated, then create it.
-					// This is to handle cases whereby we can have two types that are
-					// mapped to the same unique name -- such as in the case that we
-					// have a leafref that points to a leaf that is a union.
-					intf := goUnionInterface{
-						Name:           mtype.NativeType,
-						Types:          map[string]string{},
-						LeafPath:       field.Path(),
-						ParentReceiver: targetStruct.Name,
-					}
+			// Only if this union has more than one subtype do we generate the union;
+			// otherwise, we use that subtype directly.
+			// Also, make sure to process a union type once and only once within the struct.
+			// Even if the union has already been processed in another struct, we still need
+			// to generate the union helper with this struct as the receiver and do other
+			// processing as well. On the other hand, if the union type is used by multiple
+			// fields, we assume that it's due to a leafref and use the same union name
+			// without doing further code generation. XXX(wenbli): It's possible that it's a
```

Comment:
Is this username supposed to be here?

Reply:
Rob put his so I assumed that it's fine here as well.